### PR TITLE
removing computer object class in user list filter

### LINKED ISSUE #9353: exclude "computer" object class in user list filter
In Active Directory computers are persons, which is why they will likely show up as users.  To exclude computers the Wizard should add "(!(objectClass=computer))" to the user list filter. 

Kudos to @danimo for reporting.

### PR DESCRIPTION
**Pull request by davicente**
Fixes https://github.com/owncloud/core/issues/9353
@onwcloud/qa @blizzz

### CODE REVIEW

**Comment by PVince81** [comment]
I guess there were no unit tests affected by this change ?

File: apps/user_ldap/lib/wizard.php [hunk 1]
```diff
@@ -1221,6 +1223,8 @@ private function determineFeature($objectclasses, $attr, $dbkey, $confkey, $po =
 		if(is_array($availableFeatures)
 		   && count($availableFeatures) > 0) {
 			natcasesort($availableFeatures);
+			//Remove computer feature from the list returned by LDAP server
+			$availableFeatures = array_diff($availableFeatures, ["computer"]);
```
Comment (blizzz) [inline apps/user_ldap/lib/wizard.php]: This also removes it for the group object classes. Nevertheless, this is OK I think, as they typically are not part of that class (and would not make sense either).

**Comment by blizzz** [comment]
I found one (uncommon) edge-case:

1. before this patch, configure LDAP and  pick computer from the  objectclasses!
2. apply patch (resp. switch to this branch)
3. go  to wizard and let the filter be re-created.

The previously selected computer object class is still considered in the filter, while the exclusion is added. The user does not have the change to deselect computer anymore. 

However, this does only last until the objectClass selection is saved again. Still, it might be a nuisance.

→ Suggested solutions
1. add a check in https://github.com/owncloud/core/blob/62cad44722f0503b781ffd84ddd412495b95d8e1/apps/user_ldap/lib/wizard.php#L876 that would ignore "computer" object class
2. add an update step in appinfo/update.php that would once go through this setting and kick out 'computer' if present (ldapUserFilterObjectclass config) (recommended, keeps away overhead in biz logic).

**Comment by davicente** [comment]
@onwcloud/qa @blizzz

File: apps/user_ldap/lib/accessfactory.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ */
+
+
+namespace OCA\user_ldap\lib;
+
+
+class AccessFactory {
+
+	public static function createAccess($configPrefix) {
```
Comment (blizzz) [inline apps/user_ldap/lib/accessfactory.php]: Please don't do static. This is impossible to mock.

PHPUnit:

[spectacle w28935](https://phpunit.de/manual/current/en/test-doubles.html)

**Comment by blizzz** [comment]
Aside of that comment, the recent changes look OK.

**Comment by davicente** [comment]
> Aside of that comment, the recent changes look OK.

Done

File: apps/user_ldap/lib/accessfactory.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ */
+
+
+namespace OCA\user_ldap\lib;
+
+
+class AccessFactory {
+
+	public function __construct() {
```
Comment (blizzz) [inline apps/user_ldap/lib/accessfactory.php]: you don't need to declare the constructor when it is empty

Comment (davicente) [inline apps/user_ldap/lib/accessfactory.php]: > you don't need to declare the constructor when it is empty

Agree. Done

**Comment by davitol** [comment]
@davicente @blizzz 
It works for me, but I think it would be correct to add the & to the filter only if two object classes are selected in the combobox:

E.g:
Without selecting any object:
(!(objectClass=computer))

Selecting person as object:
(&(|(objectclass=person))(!(objectClass=computer)))

--> & should not be add due to only one argument is selected

Selecting person and user as objects:
(&(|(objectclass=person)(objectclass=user))(!(objectClass=computer)))

**Comment by davicente** [comment]
> but I think it would be better to add the ' | ' to the filter only if two object classes are selected in the combobox:

@davitol Agree, I think now it's much better and clearer to the LDAP admin, as you suggested.

**Comment by davicente** [comment]
@owncloud/qa

**Comment by davitol** [comment]
@davicente  I think you should rename Wizard.php file due to the movement of user_ldap to PSR-4

cc @PVince81

**Comment by davicente** [comment]
It's not working properly in our LDAP testing machine. I have tested manually with ldapsearh and results are not consistent. 
These 2 filters are returning empty results:
'(!(objectClass=computer))'      ->  This should return lot of results, but it doesn't
'(objectClass=computer)'
and they should be complementary. 
Rest of objectClass values work properly, for example:
'(!(objectClass=person))'   ->   return all entries without 'person' value in objectClass.
'(objectClass=person)'      ->   return all entries with 'person' value in ObjectClass.

Any idea why this happens??   Someone could test this PR with another LDAP ??

**Comment by PVince81** [comment]
Moving to 9.2 since this isn't ready and doesn't have a huge benefit (was enhancement)

**Comment by PVince81** [comment]
@jvillafanez

**Comment by jvillafanez** [comment]
Do we want to move this? It seems too much code for something an admin can do it by himself.

**Comment by PVince81** [comment]
@jvillafanez is there a way to provide a simpler fix that just adds the required filter when detecting AD ?

**Comment by jvillafanez** [comment]
The code doesn't seem to be prepared for that. I don't think it's possible to provide a clean solution without tearing apart half of the code. If we have to, better go with this.

**Comment by PVince81** [comment]
Is this something that a LDAP admin can exclude in a few minutes by editing the filter manually ?

If yes, then the complexity of this change it too high for a too small benefit.

**Comment by jvillafanez** [comment]
Having a kind of cheatsheet which includes this filter, it should be easy. I think the problem could be on our side since we need to document this properly, telling that this filter is intended for AD, plus adding some examples. I guess all depends on how much skill do we suppose to the admin.

**Comment by PVince81** [comment]
seems the complexity outweights the benefit

**Pull request closed without merging**